dask/tests/test_multiprocessing.py

```diff
@@ -194,6 +195,41 @@ def f():
     assert len(set(results)) == N
 
 
+class global_:
+    value = 0
+
+
+def proc_init():
+    global_.value = 1
+
+
+def test_process_initializer():
+    import numpy as np
+    import dask.array as da
+
```

Comment:
As this code is now, there is no problem.  ***But*** ...

... an unexpected error (related to pickling) is raised when I move the definition of `proc_init()` into `test_process_initializer()` so that it becomes a nested function.  Not sure how to fix that.

<details>
<summary>Click to see the error obtained in such a scenario:</summary>

    ============================= test session starts =============================
    platform win32 -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0 -- C:\path\to\anaconda3\envs\daskenv\python.exe
    cachedir: .pytest_cache
    rootdir: C:\path\to\git\dask, configfile: setup.cfg
    plugins: anyio-3.5.0
    collecting ... collected 21 items

    test_multiprocessing.py::test_process_initializer Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\path\to\anaconda3\envs\daskenv\lib\multiprocessing\spawn.py", line 116, in spawn_main
        exitcode = _main(fd, parent_sentinel)
      File "C:\path\to\anaconda3\envs\daskenv\lib\multiprocessing\spawn.py", line 126, in _main
        self = reduction.pickle.load(from_parent)
    EOFError: Ran out of input
    FAILED

    ================================== FAILURES ===================================
    __________________________ test_process_initializer ___________________________

        def test_process_initializer():
            import numpy as np
            import dask.array as da
        
            def proc_init():
                global_.value = 1
        
            def f(x):
                return global_.value*x
        
            global_.value = 1
            x = da.arange(9, chunks=3)
            y = x.map_blocks(f)
        
            with dask.config.set(scheduler="threading"):
                y0 = y.compute()
            expected_y = np.arange(9)
            assert_eq(y0, expected_y)  # processes saw global_.value = 1
        
            with dask.config.set(scheduler="processes"):
                y1 = y.compute()
            expected_y = np.zeros(9, dtype=np.int_)
            assert_eq(y1, expected_y)  # processes did not see global_.value = 1
        
            with dask.config.set(scheduler="processes", initializer=proc_init):
    >           y2 = y.compute()

    da         = <module 'dask.array' from 'C:\\path\\to\\git\\dask\\dask\\array\\__init__.py'>
    expected_y = array([0, 0, 0, 0, 0, 0, 0, 0, 0])
    f          = <function test_process_initializer.<locals>.f at 0x000001FDB5534F70>
    np         = <module 'numpy' from 'C:\\path\\to\\anaconda3\\envs\\daskenv\\lib\\site-packages\\numpy\\__init__.py'>
    proc_init  = <function test_process_initializer.<locals>.proc_init at 0x000001FDB697DD80>
    x          = dask.array<arange, shape=(9,), dtype=int32, chunksize=(3,), chunktype=numpy.ndarray>
    y          = dask.array<f, shape=(9,), dtype=int32, chunksize=(3,), chunktype=numpy.ndarray>
    y0         = array([0, 1, 2, 3, 4, 5, 6, 7, 8])
    y1         = array([0, 0, 0, 0, 0, 0, 0, 0, 0])

    test_multiprocessing.py:231: 
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
    ..\base.py:311: in compute
        (result,) = compute(self, traverse=False, **kwargs)
            kwargs     = {}
            self       = dask.array<f, shape=(9,), dtype=int32, chunksize=(3,), chunktype=numpy.ndarray>
    ..\base.py:599: in compute
        results = schedule(dsk, keys, **kwargs)
            args       = (dask.array<f, shape=(9,), dtype=int32, chunksize=(3,), chunktype=numpy.ndarray>,)
            collections = [dask.array<f, shape=(9,), dtype=int32, chunksize=(3,), chunktype=numpy.ndarray>]
            dsk        = {('arange-f-d7f0290c5ad4f7132c5083673cfc91af', 0): (subgraph_callable-48ca4574-0eb6-43e6-86b2-53ba4aaa4395, (functools... 3, dtype('int32'))), ('f-d7f0290c5ad4f7132c5083673cfc91af', 0): ('arange-f-d7f0290c5ad4f7132c5083673cfc91af', 0), ...}
            get        = None
            keys       = [[('f-d7f0290c5ad4f7132c5083673cfc91af', 0), ('f-d7f0290c5ad4f7132c5083673cfc91af', 1), ('f-d7f0290c5ad4f7132c5083673cfc91af', 2)]]
            kwargs     = {}
            optimize_graph = True
            postcomputes = [(<function finalize at 0x000001FDB7D46170>, ())]
            repack     = <function unpack_collections.<locals>.repack at 0x000001FDB5535990>
            schedule   = <function get at 0x000001FDB5625F30>
            scheduler  = None
            traverse   = False
            x          = dask.array<f, shape=(9,), dtype=int32, chunksize=(3,), chunktype=numpy.ndarray>
    ..\multiprocessing.py:232: in get
        result = get_async(
            chunksize  = 6
            cleanup    = True
            context    = <module 'multiprocessing' from 'C:\\path\\to\\anaconda3\\envs\\daskenv\\lib\\multiprocessing\\__init__.py'>
            dependencies = {('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0): set(), ('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 1): set(),...91af', 2): set(), ('f-d7f0290c5ad4f7132c5083673cfc91af', 0): {('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0)}, ...}
            dsk        = {('arange-f-d7f0290c5ad4f7132c5083673cfc91af', 0): (subgraph_callable-48ca4574-0eb6-43e6-86b2-53ba4aaa4395, (functools... 3, dtype('int32'))), ('f-d7f0290c5ad4f7132c5083673cfc91af', 0): ('arange-f-d7f0290c5ad4f7132c5083673cfc91af', 0), ...}
            dsk2       = {('arange-f-d7f0290c5ad4f7132c5083673cfc91af', 0): (subgraph_callable-48ca4574-0eb6-43e6-86b2-53ba4aaa4395, (functools... 3, dtype('int32'))), ('f-d7f0290c5ad4f7132c5083673cfc91af', 0): ('arange-f-d7f0290c5ad4f7132c5083673cfc91af', 0), ...}
            dsk3       = {('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0): (subgraph_callable-48ca4574-0eb6-43e6-86b2-53ba4aaa4395, (functoo..., dtype('int32'))), ('f-d7f0290c5ad4f7132c5083673cfc91af', 0): ('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0), ...}
            dumps      = functools.partial(<function dumps at 0x000001FDB561EE60>, protocol=5)
            func_dumps = None
            func_loads = None
            initializer = <function test_process_initializer.<locals>.proc_init at 0x000001FDB697DD80>
            keys       = [[('f-d7f0290c5ad4f7132c5083673cfc91af', 0), ('f-d7f0290c5ad4f7132c5083673cfc91af', 1), ('f-d7f0290c5ad4f7132c5083673cfc91af', 2)]]
            kwargs     = {}
            loads      = <built-in function loads>
            num_workers = 8
            optimize_graph = True
            pool       = <concurrent.futures.process.ProcessPoolExecutor object at 0x000001FDB5564C70>
    ..\local.py:496: in get_async
        fire_tasks(chunksize)
            _          = ()
            cache      = None
            callbacks  = ()
            chunksize  = 6
            dsk        = {('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0): (subgraph_callable-48ca4574-0eb6-43e6-86b2-53ba4aaa4395, (functoo..., dtype('int32'))), ('f-d7f0290c5ad4f7132c5083673cfc91af', 0): ('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0), ...}
            dumps      = functools.partial(<function dumps at 0x000001FDB561EE60>, protocol=5)
            fire_tasks = <function get_async.<locals>.fire_tasks at 0x000001FDB55352D0>
            get_id     = <function _process_get_id at 0x000001FDB56248B0>
            keyorder   = {('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0): 0, ('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 1): 2, ('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 2): 4, ('f-d7f0290c5ad4f7132c5083673cfc91af', 0): 1, ...}
            kwargs     = {}
            loads      = <built-in function loads>
            num_workers = 8
            pack_exception = <function pack_exception at 0x000001FDB5625E10>
            posttask_cbs = ()
            pretask_cbs = ()
            queue      = <queue.Queue object at 0x000001FDB55653C0>
            raise_exception = <function reraise at 0x000001FDB53B0E50>
            rerun_exceptions_locally = False
            result     = [[('f-d7f0290c5ad4f7132c5083673cfc91af', 0), ('f-d7f0290c5ad4f7132c5083673cfc91af', 1), ('f-d7f0290c5ad4f7132c5083673cfc91af', 2)]]
            result_flat = {('f-d7f0290c5ad4f7132c5083673cfc91af', 0), ('f-d7f0290c5ad4f7132c5083673cfc91af', 1), ('f-d7f0290c5ad4f7132c5083673cfc91af', 2)}
            results    = {('f-d7f0290c5ad4f7132c5083673cfc91af', 0), ('f-d7f0290c5ad4f7132c5083673cfc91af', 1), ('f-d7f0290c5ad4f7132c5083673cfc91af', 2)}
            started_cbs = []
            state      = {'cache': {}, 'dependencies': {('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0): set(), ('arange-f-f-d7f0290c5ad4f71...range-f-f-d7f0290c5ad4f7132c5083673cfc91af', 2): {('f-d7f0290c5ad4f7132c5083673cfc91af', 2)}}), 'finished': set(), ...}
            submit     = <bound method ProcessPoolExecutor.submit of <concurrent.futures.process.ProcessPoolExecutor object at 0x000001FDB5564C70>>
            succeeded  = False
    ..\local.py:491: in fire_tasks
        fut = submit(batch_execute_tasks, each_args)
            _          = 2
            args       = [(('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0), b'\x80\x05\x95\xd5\x03\x00\x00\x00\x00\x00\x00\x8c\x11dask.optim...in function loads>, <function _process_get_id at 0x000001FDB56248B0>, <function pack_exception at 0x000001FDB5625E10>)]
            avail_workers = 8
            chunksize  = 6
            data       = {}
            dsk        = {('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0): (subgraph_callable-48ca4574-0eb6-43e6-86b2-53ba4aaa4395, (functoo..., dtype('int32'))), ('f-d7f0290c5ad4f7132c5083673cfc91af', 0): ('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0), ...}
            dumps      = functools.partial(<function dumps at 0x000001FDB561EE60>, protocol=5)
            each_args  = [(('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0), b'\x80\x05\x95\xd5\x03\x00\x00\x00\x00\x00\x00\x8c\x11dask.optim...in function loads>, <function _process_get_id at 0x000001FDB56248B0>, <function pack_exception at 0x000001FDB5625E10>)]
            get_id     = <function _process_get_id at 0x000001FDB56248B0>
            i          = 0
            key        = ('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 2)
            loads      = <built-in function loads>
            nready     = 3
            ntasks     = 3
            num_workers = 8
            pack_exception = <function pack_exception at 0x000001FDB5625E10>
            pretask_cbs = ()
            queue      = <queue.Queue object at 0x000001FDB55653C0>
            state      = {'cache': {}, 'dependencies': {('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0): set(), ('arange-f-f-d7f0290c5ad4f71...range-f-f-d7f0290c5ad4f7132c5083673cfc91af', 2): {('f-d7f0290c5ad4f7132c5083673cfc91af', 2)}}), 'finished': set(), ...}
            submit     = <bound method ProcessPoolExecutor.submit of <concurrent.futures.process.ProcessPoolExecutor object at 0x000001FDB5564C70>>
            used_workers = 0
    ..\..\..\..\anaconda3\envs\daskenv\lib\concurrent\futures\process.py:705: in submit
        self._adjust_process_count()
            args       = ([(('arange-f-f-d7f0290c5ad4f7132c5083673cfc91af', 0), b'\x80\x05\x95\xd5\x03\x00\x00\x00\x00\x00\x00\x8c\x11dask.opti... function loads>, <function _process_get_id at 0x000001FDB56248B0>, <function pack_exception at 0x000001FDB5625E10>)],)
            f          = <Future at 0x1fdb5564e50 state=pending>
            fn         = <function batch_execute_tasks at 0x000001FDB53B0AF0>
            kwargs     = {}
            self       = <concurrent.futures.process.ProcessPoolExecutor object at 0x000001FDB5564C70>
            w          = <concurrent.futures.process._WorkItem object at 0x000001FDB5565030>
    ..\..\..\..\anaconda3\envs\daskenv\lib\concurrent\futures\process.py:683: in _adjust_process_count
        p.start()
            p          = <Process name='Process-2' parent=19112 initial>
            process_count = 0
            self       = <concurrent.futures.process.ProcessPoolExecutor object at 0x000001FDB5564C70>
    ..\..\..\..\anaconda3\envs\daskenv\lib\multiprocessing\process.py:121: in start
        self._popen = self._Popen(self)
            self       = <Process name='Process-2' parent=19112 initial>
    ..\..\..\..\anaconda3\envs\daskenv\lib\multiprocessing\context.py:224: in _Popen
        return _default_context.get_context().Process._Popen(process_obj)
            process_obj = <Process name='Process-2' parent=19112 initial>
    ..\..\..\..\anaconda3\envs\daskenv\lib\multiprocessing\context.py:327: in _Popen
        return Popen(process_obj)
            Popen      = <class 'multiprocessing.popen_spawn_win32.Popen'>
            process_obj = <Process name='Process-2' parent=19112 initial>
    ..\..\..\..\anaconda3\envs\daskenv\lib\multiprocessing\popen_spawn_win32.py:93: in __init__
        reduction.dump(process_obj, to_child)
            cmd        = '"C:\\path\\to\\anaconda3\\envs\\daskenv\\python.exe" "-c" "from multiprocessing.spawn import spawn_main; spawn_main(parent_pid=19112, pipe_handle=2296)" "--multiprocessing-fork"'
            env        = None
            hp         = 2304
            ht         = 2052
            pid        = 19548
            prep_data  = {'authkey': b'\xb160Vlsq\xd1i\xcdy\\\xe4\x87\x83eU\x1c\x1c\x03\xfc~L\x89i\xe6\xbbP\x03\xab\xec\xfb', 'dir': 'C:\\path...\to\\.p2\\pool\\plugins\\org.python.pydev.core_9.2.0.202110311311\\pysrc\\runfiles.py', 'log_to_stderr': False, ...}
            process_obj = <Process name='Process-2' parent=19112 initial>
            python_exe = 'C:\\path\\to\\anaconda3\\envs\\daskenv\\python.exe'
            rhandle    = 2296
            self       = <multiprocessing.popen_spawn_win32.Popen object at 0x000001FDB5565D50>
            tid        = 18912
            to_child   = <_io.BufferedWriter name=5>
            wfd        = 5
            whandle    = 2284
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    obj = <Process name='Process-2' parent=19112 initial>
    file = <_io.BufferedWriter name=5>, protocol = None

        def dump(obj, file, protocol=None):
            '''Replacement for pickle.dump() using ForkingPickler.'''
    >       ForkingPickler(file, protocol).dump(obj)
    E       AttributeError: Can't pickle local object 'test_process_initializer.<locals>.proc_init'

    file       = <_io.BufferedWriter name=5>
    obj        = <Process name='Process-2' parent=19112 initial>
    protocol   = None

    ..\..\..\..\anaconda3\envs\daskenv\lib\multiprocessing\reduction.py:60: AttributeError
    ============================ slowest 10 durations =============================
    1.80s call     dask/tests/test_multiprocessing.py::test_process_initializer

    (2 durations < 0.005s hidden.  Use -vv to show these durations.)
    =========================== short test summary info ===========================
    FAILED test_multiprocessing.py::test_process_initializer - AttributeError: Ca...
    ============================== 1 failed in 3.03s ==============================

</details>